recipe/0005-CMake-Explicitly-link-with-CCCL-standalone-or-CTK-10.patch

```diff
@@ -0,0 +1,63 @@
+From 15b7012f54793f3cf807501453a61b6a6c1fb3c3 Mon Sep 17 00:00:00 2001
+From: Philip Hyunsu Cho <chohyu01@cs.washington.edu>
+Date: Tue, 23 Jul 2024 18:42:54 -0700
+Subject: [PATCH] [CMake] Explicitly link with CCCL (standalone or CTK)
+ (#10624)
+
+---
+ CMakeLists.txt    | 18 ++++++++++++++++++
+ cmake/Utils.cmake |  6 ++++--
+ 2 files changed, 22 insertions(+), 2 deletions(-)
+
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index dc0f7e654..aa078405f 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -224,6 +224,24 @@ if(USE_CUDA)
+   add_subdirectory(${PROJECT_SOURCE_DIR}/gputreeshap)
+ 
+   find_package(CUDAToolkit REQUIRED)
++  find_package(CCCL CONFIG)
++  if(NOT CCCL_FOUND)
++    message(STATUS "Standalone CCCL not found. Attempting to use CCCL from CUDA Toolkit...")
++    find_package(CCCL CONFIG
++      HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++    if(NOT CCCL_FOUND)
++      message(STATUS "Could not locate CCCL from CUDA Toolkit. Using Thrust and CUB from CUDA Toolkit...")
++      find_package(libcudacxx CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(CUB CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(Thrust CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      thrust_create_target(Thrust HOST CPP DEVICE CUDA)
++      add_library(CCCL::CCCL INTERFACE IMPORTED GLOBAL)
++      target_link_libraries(CCCL::CCCL INTERFACE libcudacxx::libcudacxx CUB::CUB Thrust)
++    endif()
++  endif()
+ endif()
+ 
+ if(FORCE_COLORED_OUTPUT AND (CMAKE_GENERATOR STREQUAL "Ninja") AND
+diff --git a/cmake/Utils.cmake b/cmake/Utils.cmake
+index 8659df7c3..443b77c7b 100644
+--- a/cmake/Utils.cmake
++++ b/cmake/Utils.cmake
+@@ -108,11 +108,13 @@ function(xgboost_set_cuda_flags target)
+     target_compile_definitions(${target} PRIVATE -DXGBOOST_USE_NVTX=1)
+   endif()
+ 
++  target_link_libraries(${target} PRIVATE CCCL::CCCL)
+   target_compile_definitions(${target} PRIVATE -DXGBOOST_USE_CUDA=1)
+   target_include_directories(
+     ${target} PRIVATE
+-    ${xgboost_SOURCE_DIR}/gputreeshap
+-    ${xgboost_SOURCE_DIR}/rabit/include
++    ${xgboost_SOURCE_DIR}/gputreeshap)
++  target_include_directories(
++    ${target} SYSTEM PRIVATE
```

Comment:
Also wonder if using `AFTER` might help (based on [this doc]( https://cmake.org/cmake/help/v3.30/command/target_include_directories.html )):

> By using `AFTER` or `BEFORE` explicitly, you can select between appending and prepending, independent of the default.

```suggestion
+    ${target} SYSTEM AFTER PRIVATE
```

Reply:
Actually `SYSTEM` may already do this for us. Later in the same doc:

> Additionally, system include directories are searched after normal include directories regardless of the order specified.

Reply:
> system include directories are searched after normal include directories regardless of the order specified.

So not sure if `AFTER` would have an effect. I can give it a try.

Reply:
Maybe we can try `AFTER` instead of `SYSTEM`?

Am guessing `SYSTEM` translates to `-isystem`, which might be resulting in some of the errors we are now seeing on CI

```suggestion
+    ${target} AFTER PRIVATE
```

Reply:
Tried adding this to see if it would help. Though happy to try other things if there are better options